src/main/java/io/cdap/plugin/gcp/bigquery/relational/BigQueryRelation.java

```diff
@@ -1,117 +1,409 @@
 import io.cdap.cdap.etl.api.relational.Relation;
+import io.cdap.plugin.gcp.bigquery.sqlengine.BigQuerySQLDataset;
+import io.cdap.plugin.gcp.bigquery.sqlengine.builder.BigQueryDeduplicateSQLBuilder;
+import io.cdap.plugin.gcp.bigquery.sqlengine.builder.BigQueryGroupBySQLBuilder;
+import io.cdap.plugin.gcp.bigquery.sqlengine.builder.BigQueryNestedSelectSQLBuilder;
+import io.cdap.plugin.gcp.bigquery.sqlengine.builder.BigQuerySelectSQLBuilder;
 
-import java.util.HashMap;
+import java.util.Collection;
+import java.util.LinkedHashMap;
 import java.util.Map;
 import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
 
 /**
  * An implementation of {@link Relation} designed to operate on BigQuery.
  */
 public class BigQueryRelation implements Relation {
-  private final boolean isValid;
-  private final String validationError;
+  private static final SQLExpressionFactory factory = new SQLExpressionFactory();
 
-  private final Map<String, Expression> columns;
-  private final Expression filter;
-  private final Relation baseRelation;
   private final String datasetName;
+  private final Set<String> columns;
+  private final BigQueryRelation parent;
+  private Map<String, BigQuerySQLDataset> sourceDatasets;
+  private final Supplier<String> expressionSupplier;
 
   /**
-   * Initializes a new {@link BigQueryRelation} using the set of columns specified.
-   * @param columnSet set of strings indicating column names in the relation.
+   * Gets a new BigQueryRelation instance
+   *
+   * @param datasetName source dataset name
+   * @param columnNames column names to use when initializing this relation.
+   * @return new BigQueryRelation instance for this table.
    */
-  public BigQueryRelation(String datasetName, Set<String> columnSet) {
-    SQLExpressionFactory factory = new SQLExpressionFactory();
+  public static BigQueryRelation getInstance(String datasetName,
+                                             Set<String> columnNames) {
+    return new BigQueryRelation(datasetName, columnNames);
+  }
 
-    Map<String, Expression> columns = new HashMap<>();
-    for (String column: columnSet) {
-      columns.put(column, factory.compile(column));
-    }
+  @VisibleForTesting
+  protected BigQueryRelation(String datasetName,
+                             Set<String> columns) {
     this.datasetName = datasetName;
     this.columns = columns;
-    this.isValid = true;
-    this.validationError = null;
-    this.filter = null;
-    this.baseRelation = null;
-  }
-
-  private BigQueryRelation(String datasetName,
-                           Map<String, Expression> columns,
-                           Expression filter,
-                           Relation baseRelation) {
-    this(datasetName, columns, filter, baseRelation, true, null);
+    this.parent = null;
+    this.expressionSupplier = () -> {
+      BigQuerySQLDataset sourceDataset = sourceDatasets.get(datasetName);
```

Comment:
Please handle the case when it's null. If for whatever reason it get out of sync it would be hard to diagnose with just NPE